src/Components/Components/src/Routing/Router.cs

```diff
@@ -194,51 +195,58 @@ internal virtual void Refresh(bool isNavigationIntercepted)
 
         private async ValueTask<bool> RunOnNavigateAsync(string path, Task previousOnNavigate)
         {
-            // If this router instance does not provide an OnNavigateAsync parameter
-            // then we render the component associated with the route as per usual.
             if (!OnNavigateAsync.HasDelegate)
             {
                 return true;
             }
 
-            // If we've already invoked a task and stored its CTS, then
-            // cancel that existing CTS.
+            // Cancel the CTS instead of disposing it, since disposing does not
+            // actually cancel and can cause unintended Object Disposed Exceptions.
+            // This effectivelly cancels the previously running task and completes it.
             _onNavigateCts?.Cancel();
-            // Then make sure that the task has been completed cancelled or
-            // completed before continuing with the execution of this current task.
+            // Then make sure that the task has been completely cancelled or completed
+            // before starting the next one. This avoid race conditions where the cancellation
+            // for the previous task was set but not fully completed by the time we get to this
+            // invocation.
             await previousOnNavigate;
 
-            // Create a new cancellation token source for this instance
             _onNavigateCts = new CancellationTokenSource();
             var navigateContext = new NavigationContext(path, _onNavigateCts.Token);
 
-            // Create a cancellation task based on the cancellation token
-            // associated with the current running task.
-            var cancellationTcs = new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously);
-            navigateContext.CancellationToken.Register(state =>
-                ((TaskCompletionSource)state).SetResult(), cancellationTcs);
-
             var task = OnNavigateAsync.InvokeAsync(navigateContext);
 
-            // If the user provided a Navigating render fragment, then show it.
-            if (Navigating != null && task.Status != TaskStatus.RanToCompletion)
+            try
+            {
+                if (Navigating != null)
+                {
+                    _renderHandle.Render(Navigating);
+                }
+                await task;
+                return true;
+            }
+            catch (OperationCanceledException e)
+            {
+                if (task.IsCanceled && !navigateContext.CancellationToken.IsCancellationRequested)
+                {
+                    throw new InvalidOperationException("OnNavigateAsync can only be cancelled via he NavigateContext.CancellationToken.", e);
```

Comment:
Reading the other discussion, it makes sense to me why we want this rethrown and turned into an InvalidOperationException instead of swallowed. What I don't understand is why other non-OperationCanceledExceptions aren't rethrown the same way and are instead dispatched to the renderer. Why not surface all exceptions that aren't swallowed the same way?